### PR TITLE
[AIRFLOW-4549] skipped tasks should be ok for wait_for_downstream

### PR DESCRIPTION
**Pull request by dima-asana**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4549

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
this PR updates wait_for_downstream dependency to be satisfied for tasks that are skipped.  This is desirable because 
(a) depends_on_past, which this is interacts with, allows skipped tasks
(b) documentation for wait_for_downstream mentions it is to avoid resource conflicts, which are de facto avoided when a task is skipped

See e.g. discussion at http://mail-archives.apache.org/mod_mbox/airflow-dev/201609.mbox/%3CCAHEEp7UTGPjVKgww9_9N5FUpNU+PSKF3RmBvXUGK5dxb6bhjbw@mail.gmail.com%3E 

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
DagRunTest.test_wait_for_downstream

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: tests/models/test_dagrun.py
Comment (dima-asana) [inline tests/models/test_dagrun.py:564]: I'm not really sure where to put these integration-like tests.  They only check functionality on task instances.  But they require having DagRuns for the tasks because of how `previous_ti` works (https://github.com/apache/airflow/blob/e1c1a8dad0e01a7baa50ebe02e748429d33241fd/airflow/models/taskinstance.py#L533-L540).  They also require something to actually try running the tasks (I used dag.run).  And the exception they raise is because of scheduler deadlock as expected.  So I could rationalize putting this here, SchedulerJobTest, or TaskInstanceTest.  Let me know if someone more familiar with the code has a preference.

I'm also not sure they're strictly necessary.  I think this code is reasonably unit tested through the combination of https://github.com/apache/airflow/blob/e1c1a8dad0e01a7baa50ebe02e748429d33241fd/tests/ti_deps/deps/test_prev_dagrun_dep.py and https://github.com/apache/airflow/blob/5acc221a0993296b00430626de107515acb9c2d4/tests/models/test_taskinstance.py#L519-L545.  At the same time, it might be more straightforward to follow this example in an actual Dag like here, and the task_instance.run() part of this is only tested in accidentally tangent unit tests.

File: airflow/models/taskinstance.py
Comment (dima-asana) [inline airflow/models/taskinstance.py:556]: (this is the only line of non-test code that is changed -- the rest is either testing or documentation)

File: tests/models/test_taskinstance.py
Comment (dima-asana) [inline tests/models/test_taskinstance.py]: I replaced this explicit example and moved it to an integration test above.  The functionality here gets tested deeper in by https://github.com/apache/airflow/blob/e1c1a8dad0e01a7baa50ebe02e748429d33241fd/tests/ti_deps/deps/test_prev_dagrun_dep.py#L35-L67

File: tests/models/test_dagrun.py
Comment (milton0825) [inline tests/models/test_dagrun.py:565]: Love the test.

**Review by milton0825 — APPROVED** [review]
Looks good

**Comment by OmerJog** [comment]
@dima-asana can you resolve conflict?
@ashb @kaxil i think this PR is important for 1.10.4

**Review by ashb — COMMENTED** [review]
We should probably mention this in the UPDATING.md -- it's a change in behaviour.

**Comment by dima-asana** [comment]
> @dima-asana can you resolve conflict?

done

> We should probably mention this in the UPDATING.md -- it's a change in behaviour.

done

**Comment by shaikshakeel** [comment]
Awesome, actually looking for this. and one more suggestion why can't we have state like `done`. because in my use-case I don't care about past tasks whether the task is a success or failure. just have to be finished.  `state not in enqueued or running`.

Any thoughts? am I missing anything?

**Comment by OmerJog** [comment]
@dima-asana @milton0825 a rebase is needed

**Comment by OmerJog** [comment]
@dima-asana any progress with rebasing?

**Comment by teddyhartanto** [comment]
Hi, any update on this? This is desperately needed. It doesn't make sense that we regard `Status.SKIPPED` and `Status.SUCCESS` as "successful" state in the `depends_on_past` check, but for `wait_for_downstream`, only `Status.SUCCESS` is regarded as "successful" state. This needs a fix.

And because it seems that @dima-asana has gone missing, should we start a new PR? I can help.

**Comment by potiuk** [comment]
@TeddyHartanto -> feel free to start a new PR, happy to help with review

**Comment by kaxil** [comment]
I have resolved the conflicts @potiuk and @TeddyHartanto .

I think we can merge it as it is. Already has an approval, any objections, comment @potiuk or @ashb ?

Although, let's wait for the CI results.

**Comment by potiuk** [comment]
No objections. Just static checks failure :(

**Comment by teddyhartanto** [comment]
@kaxil no objection too. Just the static checks failure as mentioned by @potiuk

**Comment by teddyhartanto** [comment]
@kaxil the build is still failing. Do you mind taking a look at it? It's coming from `tests/models/test_dagrun.py`:
```from tests.executors.test_executor import TestExecutor```

There's no package `tests.executors.test_executor` nor the class `TestExecutor`

**Comment by kaxil** [comment]
@TeddyHartanto Some of the tests are failing, I think it would be better to start a new PR, want to give it a shot?

**Comment by TeddyHartanto** [comment]
@kaxil Sure. On it!

**Comment by kaxil** [comment]
> @kaxil Sure. On it!

Awesome, thanks

**Pull request closed without merging**
